bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,408 @@
+	///
+	/// Inspired by the `frame_system::Pallet::Events` storage value
+	#[pallet::storage]
+	#[pallet::unbounded]
+	pub(super) type Messages<T: Config> = StorageValue<_, Vec<OutboundMessage>, ValueQuery>;
+
+	/// Hashes of the ABI-encoded messages in the [`Messages`] storage value. Used to generate a
+	/// merkle root during `on_finalize`. This storage value is killed in
+	/// `on_initialize`, so should never go into block PoV.
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn message_leaves)]
+	pub(super) type MessageLeaves<T: Config> = StorageValue<_, Vec<H256>, ValueQuery>;
+
+	/// The current nonce for the messages
+	#[pallet::storage]
+	pub type Nonce<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	#[pallet::getter(fn operating_mode)]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	/// Pending orders to relay
+	#[pallet::storage]
+	pub type PendingOrders<T: Config> =
+		StorageMap<_, Identity, u64, PendingOrder<BlockNumberFor<T>>, OptionQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		T::AccountId: AsRef<[u8]>,
+	{
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			// Remove storage from previous block
+			Messages::<T>::kill();
+			MessageLeaves::<T>::kill();
+			// Reserve some weight for the `on_finalize` handler
+			T::WeightInfo::commit()
+		}
+
+		fn on_finalize(_: BlockNumberFor<T>) {
+			Self::commit();
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> where T::AccountId: From<[u8; 32]> {
+		/// Halt or resume all pallet operations. May only be called by root.
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operating_mode(
+			origin: OriginFor<T>,
+			mode: BasicOperatingMode,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			OperatingMode::<T>::put(mode);
+			Self::deposit_event(Event::OperatingModeChanged { mode });
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::submit_delivery_receipt())]
+		pub fn submit_delivery_receipt(
+			origin: OriginFor<T>,
+			event: Box<EventProof>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			ensure!(!Self::operating_mode().is_halted(), Error::<T>::Halted);
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&event.event_log, &event.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			let receipt = MessageReceiptOf::<T>::try_from(&event.event_log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			// Verify that the message was submitted from the known Gateway contract
+			ensure!(T::GatewayAddress::get() == receipt.gateway, Error::<T>::InvalidGateway);
+
+			let nonce = receipt.nonce;
+
+			let order = <PendingOrders<T>>::get(nonce).ok_or(Error::<T>::InvalidPendingNonce)?;
+
+			// No fee for governance order
```

Comment:
@yrong rather remove comment, to imply that remote fees are always optional

```suggestion
```

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/b13e3f26b31b905eae1d67ac6574129d67d07a07